File: camera/camera-camera2/src/main/java/androidx/camera/camera2/internal/compat/workaround/UseFlashModeTorchFor3aUpdate.java
/*
 * Copyright 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.camera.camera2.internal.compat.workaround;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.camera.camera2.internal.compat.quirk.TorchFlashRequiredFor3aUpdateQuirk;
import androidx.camera.core.Logger;
import androidx.camera.core.impl.Quirks;

/**
 * Workaround to use torch as flash.
 *
 * @see TorchFlashRequiredFor3aUpdateQuirk
 */
@RequiresApi(21) // TODO(b/200306659): Remove and replace with annotation on package-info.java
public class UseFlashModeTorchFor3aUpdate {
    private static final String TAG = "UseFlashModeTorchFor3aUpdate";

    @Nullable
    private final TorchFlashRequiredFor3aUpdateQuirk mTorchFlashRequiredFor3AUpdateQuirk;

    public UseFlashModeTorchFor3aUpdate(@NonNull Quirks quirks) {
        mTorchFlashRequiredFor3AUpdateQuirk = quirks.get(TorchFlashRequiredFor3aUpdateQuirk.class);
    }

    /** Returns if torch should be used as flash. */
    public boolean shouldUseFlashModeTorch() {
        boolean shouldUse = mTorchFlashRequiredFor3AUpdateQuirk != null
                && mTorchFlashRequiredFor3AUpdateQuirk.isFlashModeTorchRequired();
        Logger.d(TAG, "shouldUseFlashModeTorch: " + shouldUse);
        return shouldUse;
    }
}

File: camera/camera-camera2-pipe-integration/src/main/java/androidx/camera/camera2/pipe/integration/compat/workaround/UseFlashModeTorchFor3aUpdate.kt
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.camera.camera2.pipe.integration.compat.workaround

import android.hardware.camera2.CaptureRequest
import androidx.camera.camera2.pipe.integration.compat.quirk.CameraQuirks
import androidx.camera.camera2.pipe.integration.compat.quirk.TorchFlashRequiredFor3aUpdateQuirk
import dagger.Module
import dagger.Provides

/**
 * Workaround to use [CaptureRequest.FLASH_MODE_TORCH] for 3A operation.
 *
 * @see TorchFlashRequiredFor3aUpdateQuirk
 */
public interface UseFlashModeTorchFor3aUpdate {
    public fun shouldUseFlashModeTorch(): Boolean

    @Module
    public abstract class Bindings {
        public companion object {
            @Provides
            public fun provideUseFlashModeTorchFor3aUpdate(
                cameraQuirks: CameraQuirks
            ): UseFlashModeTorchFor3aUpdate =
                if (cameraQuirks.quirks.contains(TorchFlashRequiredFor3aUpdateQuirk::class.java))
                    UseFlashModeTorchFor3aUpdateImpl
                else NotUseFlashModeTorchFor3aUpdate
        }
    }
}

public object UseFlashModeTorchFor3aUpdateImpl : UseFlashModeTorchFor3aUpdate {
    /** Returns true for torch should be used as flash. */
    override fun shouldUseFlashModeTorch(): Boolean = true
}

public object NotUseFlashModeTorchFor3aUpdate : UseFlashModeTorchFor3aUpdate {
    override fun shouldUseFlashModeTorch(): Boolean = false
}
